### DISCUSSION TITLE
Are you able to add custom .gir files to generate bindings for when built?

### DISCUSSION THREAD

**Discussion by foxyfocus** [gircore/gir.core#1105] [General]
How would I go about doing this? I am trying to add Gtk4LayerShell-1.0.gir, [https://github.com/wmww/gtk4-layer-shell](url)? I am on Linux, by the way.

**Comment by badcel** [comment]
Hi,

this is actively worked on by @adamreeve (See #1078).

If the work is finished there will be a nuget tool which can be used to create new bindings. 

As this tool has not yet landed there are two ways going forward:
- If the desired library is already part of the GNOME flatpak SDK it is easy to add to Gir.core directly and can be maintained here directly.
- Otherwise you need to create your own binary: 
  - Checkout the latest GirCore release (including git submodules).
  - Add the Gir file of your desired library into the Gir platform folder it is made for (under "ext/gir-files")
  - Add a new folder into the `Libs` folder matching the name of your Gir file exactly.
  - Add some minimal project structure (compare [GtkSource-5](https://github.com/gircore/gir.core/tree/main/src/Libs/GtkSource-5))
   - Add the name of the Gir file to `GenerateLibs.fsx` either as a commandline arg or inside the script
   - Call the script to generate the whole set of bindings including your library. 
   - As you checked out a GirCore release the binary will be compatible to the official binaries.

Please be aware that your library can have some API which is not yet supported by the generator and it would get skipped or the generation could even fail. In this case bug fixes are very welcome here to improve the code.

This happens regularly for the already supported libraries as they progress. It is not seldom as the generator only supports things which are needed. Supporting every kind of possible parameter combination would be extremely time consuming before any release would be possible.

If you have success and like what you were doing: Any helping hand is welcome here. You can focus on things important to improve the binding for the LayerShell library for example. This would benefit all other libraries, too.

**Comment by foxyfocus** [comment]
Okay, thanks! However, I have a question: do you have to modify GirCore.Libs.slnf as well, in order to run dotnet build GirCore.Libs.slnf? I  tried to and I received this error: 
![Screenshot_20240804_192345](https://github.com/user-attachments/assets/b670fd24-72b1-4af6-9af4-2f7a66cf800d) I also added the project using the dotnet sln command but to no avail.

**Reply by adamreeve** [reply]
You should be able to build your project by itself without needing to add it to the .sln or .slnf file:
```
dotnet build Libs/Gtk4LayerShell-1.0
```

Your error is saying that you've added your project to `GirCore.Libs.slnf` but that it isn't in the `GirCore.sln` solution. The `.slnf` file is a "solution filter" that specifies a set of projects to include from a solution file. So if you want to be able to build it as part of building `GirCore.Libs.slnf`, you'll need to also add it to the solution:
```
dotnet sln GirCore.sln add Libs/Gtk4LayerShell-1.0
```

**Reply by foxyfocus** [reply]
Okay, thanks! I managed to get it working.

**Reply by MicrogamerCz** [reply]
> Okay, thanks! I managed to get it working.

Hi, I'm trying to do the same. I can build it, but it's not working. Have you had to do some adjustments for the functionality?

**Reply by adamreeve** [reply]
> but it's not working

I don't think this should have changed. Can you please provide some details on what you're doing and the error message you get

**Reply by MicrogamerCz** [reply]
So, this is the exception I'm getting whenever I try to run a basic Gtk4 program (based on Gtk4 Layer Shell sample):

```
UnhandledException - unhandled exception: System.DllNotFoundException: Unable to load shared library 'Gtk4LayerShell' or one of its dependencies. In order to help diagnose loading problems, consider using a tool like strace. If you're using glibc, consider setting the LD_DEBUG environment variable: 
/usr/share/dotnet/shared/Microsoft.NETCore.App/8.0.8/Gtk4LayerShell.so: cannot open shared object file: No such file or directory
/home/micro/gir.core/src/Samples/Gtk-4.0/Window/bin/Debug/net8.0/Gtk4LayerShell.so: cannot open shared object file: No such file or directory
/usr/share/dotnet/shared/Microsoft.NETCore.App/8.0.8/libGtk4LayerShell.so: cannot open shared object file: No such file or directory
/home/micro/gir.core/src/Samples/Gtk-4.0/Window/bin/Debug/net8.0/libGtk4LayerShell.so: cannot open shared object file: No such file or directory
/usr/share/dotnet/shared/Microsoft.NETCore.App/8.0.8/Gtk4LayerShell: cannot open shared object file: No such file or directory
/home/micro/gir.core/src/Samples/Gtk-4.0/Window/bin/Debug/net8.0/Gtk4LayerShell: cannot open shared object file: No such file or directory
/usr/share/dotnet/shared/Microsoft.NETCore.App/8.0.8/libGtk4LayerShell: cannot open shared object file: No such file or directory
/home/micro/gir.core/src/Samples/Gtk-4.0/Window/bin/Debug/net8.0/libGtk4LayerShell: cannot open shared object file: No such file or directory

   at Gtk4LayerShell.Internal.Functions.InitForWindow(IntPtr window)
   at Gtk4LayerShell.Functions.InitForWindow(Window window) in /home/micro/gir.core/src/Libs/Gtk4LayerShell-1.0/Public/Functions.Generated.cs:line 151
   at Program.<>c.<<Main>$>b__0_0(Application sender, EventArgs args) in /home/micro/gir.core/src/Samples/Gtk-4.0/Window/Program.cs:line 6
   at GObject.Signal`1.<>c__DisplayClass11_0.<Connect>b__0(Value returnValue, Value[] parameters) in /home/micro/gir.core/src/Libs/GObject-2.0/Public/SignalTSender.cs:line 41
   at GObject.Closure.InternalCallback(IntPtr closure, IntPtr returnValuePtr, UInt32 nParamValues, IntPtr paramValuesData, IntPtr invocationHint, IntPtr userData) in /home/micro/gir.core/src/Libs/GObject-2.0/Public/Closure.cs:line 45
```

This is without any adjustments, just following the instructions here. I can actually run the program successfully if I change Library field in Internal.ImportResolver.Generated.cs to "libgtk4-layer-shell.so.0", but then nothing actually happens, it just creates a Gtk4 window with a text

**Reply by badcel** [reply]
Do you call `GtkLayerShell.Module.Initialize`?

See [here](https://github.com/gircore/gir.core/blob/main/docs/docs/faq.md#dll-not-found-exception).

For example to use WebKit you need to do the same for WebKit, see [here](https://github.com/gircore/gir.core/blob/acc83c6c64dd6aa72fb4c87a4d5145a9a2669451/src/Samples/WebKit-6.0/JavascriptCall/Program.cs#L3).

**Reply by MicrogamerCz** [reply]
I didn't, I forgot about that, it can run now fine, but the functionality doesn't work still

```
Gtk4LayerShell.Module.Initialize();

var application = Gtk.Application.New("org.gir.core", Gio.ApplicationFlags.FlagsNone);
application.OnActivate += (sender, args) =>
{
    var window = Gtk.ApplicationWindow.New((Gtk.Application)sender);
    Gtk4LayerShell.Functions.InitForWindow(window);

    Gtk4LayerShell.Functions.AutoExclusiveZoneEnable(window);
    Gtk4LayerShell.Functions.SetMargin(window, Gtk4LayerShell.Edge.Top, 10);
    Gtk4LayerShell.Functions.SetAnchor(window, Gtk4LayerShell.Edge.Top, true);

    Gtk.Widget text = Gtk.Label.New("Hello, world");
    window.SetChild(text);
    window.Show();
};
return application.RunWithSynchronizationContext(null);
```

Just the Window sample with example from Gtk4 Layer Shell, but it's still a regular window instead of being an overlay

**Reply by badcel** [reply]
I don't know anything about GtkLayerShell but if there are no exceptions / errors in the console window it should work.

Nevertheless I can see that your code is not exactly matching the [sample](https://github.com/wmww/gtk4-layer-shell/blob/main/examples/simple-example.vala). You are missing margins and you call `Show` instead of `Present`. Perhaps this is the problem?

**Reply by MicrogamerCz** [reply]
Sorry for the late answer. No, it's not the problem. Both Show and Present end up with the same result. Neither with both margins, that's more the the spacing between the border of the layer and the actual application itself

**Reply by badcel** [reply]
In this case I would recommend compiling the layer sharp sample c code to verify the problem is not related to your system.

**Reply by MicrogamerCz** [reply]
I built and ran them, they all work. C sample works as well as Vala and Python one

**Reply by badcel** [reply]
That's unfortunate.

Do you have your new code online to verify?

Does the version of the generator you used to generate the code match the version of the nuget packages you referenced? I'm not sure if this could be a problem, it's only a wild guess.

**Reply by MicrogamerCz** [reply]
I used project reference over package. I did a quick fork adding Gtk4LayerShell and changing Gtk-4.0/Window sample to use Gtk4LayerShell.

https://github.com/MicrogamerCz/gir.core

The only manual change after cloning is copying the gir file from /usr/share (cp /usr/share/gir-1.0/Gtk4LayerShell-1.0.gir <path>/ext/gir-files/linux/Gtk4LayerShell-1.0.gir)

**Reply by badcel** [reply]
Thanks i will try it out myself.

**Reply by badcel** [reply]
Ah sorry I can't test. I tried it but it failed with the following messages:

> ** (Window:16801): WARNING **: 20:09:16.323: it appears your Wayland compositor does not support the Layer Shell protocol
** (Window:16801): WARNING **: 20:09:16.323: Failed to initialize layer surface, it appears your Wayland compositor doesn't support Layer Shell
** (Window:16801): WARNING **: 20:09:16.324: GtkWindow is not a layer surface. Make sure you called gtk_layer_init_for_window ()
** (Window:16801): WARNING **: 20:09:16.324: GtkWindow is not a layer surface. Make sure you called gtk_layer_init_for_window ()

After looking into the documentation: 
>Layer shell is not supported on:
Gnome-on-Wayland
Any X11 desktop

It is clear that it can't work as I only use gnome. If you use a supported wayland compositor the only other thing I came across was that there are two versions of the package one for GTK3 and one for GTK4. From your copy command it looks like you use the GTK4 one which would be correct as GTK3 is not supported by GirCore.

If this all matches up I really don't know what is happening if there is nothing in your terminal window. Especially as the API of the Library is very simple and includes nothing special.

If dotnet could not call into the library there would be some kind of error. This means the API it wants to call is found and called. If you step into the code you see how simple those calls are. There is really nothing special.

The only other thing which comes to my mind is that you actaully run the regular window code, without the added GtkLayerShell api calls. This would explain your behavior: No errors and a simple window.

If you add this add the beginning of your program:
```CS
var myWriter = new ConsoleTraceListener();
Trace.Listeners.Add(myWriter);
```

and run with `$ dotnet run --configuration Debug` you should at least see some Debug-Output from GirCore. But I'm not sure it will help.

**Reply by MicrogamerCz** [reply]
I use Hyprland compositor, so there's no compatibility problem on my side and with some other Gtk4LS methods it seems as if it's working, but not actually modifying the window

![image](https://github.com/user-attachments/assets/eefaf658-ddc3-4cb2-a7bf-27d97344275d)

Gtk4LayerShell.Functions.IsSupported() returns true, IsLayerWindow for the window also returns true, edge anchor is also correct, but the exclusive zone doesn't work and neither the window. Handles from the trace listener are called from window and label constructors, but nothing else. And there's also nothing before the HarfBuzz errors, that's probably related to GTK4 and not GTK4LS

**Reply by badcel** [reply]
I wonder what is the difference to the C program? Why is it behaving differently?

**Reply by MicrogamerCz** [reply]
Not just C, it works in python and vala as well. I'll try to generate new Gtk3 and Gtk3LS bindings to see if that works.

**Reply by badcel** [reply]
The GirCore generator does not support GTK3.

There are APIs which are not very binding friendly and probably cant be generated. This is why GTK3 was removed some years ago. And it will probably pretty tedious to try as you would need to generate the whole stack which is several bindings like, GDK, too.

Another idea would be to write a new program without GirCore at all and just do exactly the same as the C program: So manually define the DLLImports without all the stuff GirCore adds like memory management classes and so on. Then call all this functions and see what happens.

As your demo program is pretty short it should not be too much work.

Which distribution are you using? Are your system binaries build in some special way? Perhaps there is some very low level difference so the call gets lost. GirCore does not handle C calling conventions in any special way.

Which system architecture are you using? Is it different from x64? What kind of device are you using?

**Reply by MicrogamerCz** [reply]
Ah, that's unfortunate. Gtk3 is quite easy to work with in Vala.

Anyway, I finally had some time, so I wrote it manually and it still doesn't work. I'm not entirely sure, but the problem might be related to gtk_layer_init_for_window. The last two things I can try is building and installing from source instead of a package and rewriting some parts of gtk_layer_init_for_window in C#

I use arch, so the packages are most likely built directly from the source code without changes and I have a normal laptop (x86-64)

**Comment by badcel** [comment]
This is super strange especially if GTK works like expected but layershell does not.

If it has the same wrong behaviour if you write the C# code  matching the C code without GirCore at all I would say GirCore is not to blame here.

The question is what is going wrong? You could try to check which binaries are called during runtime via [ldd](https://unix.stackexchange.com/questions/120015/how-to-find-out-the-dynamic-libraries-executables-loads-when-run).

**Reply by MicrogamerCz** [reply]
It didn't show any GTK-related libraries, only whatever is necessary for running C# apps on Linux. What I found that works is writing it with... GTK 3. Not exactly a great idea, but it seems I will have to use the older Gtk3 bindings or make a new ones at the cost of stability or maintainability.

I'm glad it's not a gir.core bug

**Reply by badcel** [reply]
It has to show GTK related libraries if you call something like `ldd dotnet run Window.csproj`.

**Reply by MicrogamerCz** [reply]
No, it doesn't.
![obrazek](https://github.com/user-attachments/assets/42b0877e-43ce-4b4a-a31e-690362183058)

And even using ldd on a published executable it doesn't show anything
![obrazek](https://github.com/user-attachments/assets/9c84eb57-65ff-41a7-8693-bbd64c8e44a4)

Maybe there is just some issue with Interop of Gtk4 and Gtk4LayerShell. Writing my program with GtkSharp (Gtk3) and manual interop of GtkLayerShell ends up working just fine.

**Reply by badcel** [reply]
Ah sorry it was the wrong command: `LD_DEBUG=libs ./Katana 2>&1 | grep gtk`
